add webpack 5 announcement (#4057)
webpack/webpack.js.org: src/content/blog/2020-10-10-webpack-5-release.md

---
title: Webpack 5 release (2020-10-10)
sort: -202010100
contributors:
    - sokra
---

webpack 4 was released in February 2018.

> Since then we shipped a lot features without breaking changes.

Reviewer: ```suggestion
Since then we shipped a lot of features without breaking changes.
```

We know that people dislike major changes with breaking changes.

> Especially with webpack, which people usually only touch twice a year and the remaining time it "just works".

Reviewer: ```suggestion
Especially with webpack, which people usually only touch twice a year, and the remaining time it "just works".
```

But shipping features without breaking changes also has a cost:
We can't do major API or architectural improvements.

> So from time to time there is a point where the difficulties pile up and we are forced to do breaking changes to not mess everything up.

Reviewer: ```suggestion
So from time to time, there is a point where the difficulties pile up and we are forced to do breaking changes to not mess everything up.
```

That's the time for a new major version.

> So webpack 5 contains these architectural improvements and the features that where not possible to implement without them.

Reviewer: ```suggestion
So webpack 5 contains these architectural improvements and the features that were not possible to implement without them.
```

Author: Isn't that the past from of `be`?

Reviewer: `where` refers to physical or metaphorical location while `were` is a verb, it’s the past tense form of `are`.

The major version was also the chance to revise some of the defaults and to align with proposals and specifications that come up in the meantime.

So today (2020-10-10) webpack 5.0.0 is released, but this doesn't mean it's done, bugfree or even feature-complete.
As with webpack 4 we continue development by fixing problems and adding features.
In the next days there will probably a lot bugfixes. Features will come later.

> So what does the release mean?

Reviewer 2: Can we make this a sub header?

It means we finished doing breaking changes.

> Many refactorings has been done to up-level the architecture and create a good base for future features (and current features).

Reviewer: ```suggestion
Many refactorings have been done to up-level the architecture and create a good base for future features (and current features).
```

> So when is the time to upgrade?

Reviewer 2: Also subheader

It depends. There is a good chance that upgrading fails and you would need to give it a second or 3rd try.
If you are open to that, try to upgrade now and provide feedback to webpack, plugins and loaders.
We are eager to fix those problems. Someone has to start and you would be one of the first ones benefiting from it.

> ## Sponsoring situation

Reviewer 2: s/situation/Update

webpack is fully based upon [sponsoring](https://opencollective.com/webpack).

> It's not tied to (and payed by) a big company like some other Open Source projects.

Reviewer: ```suggestion
It's not tied to (and paid by) a big company like some other Open Source projects.
```

99% of the earnings from sponsoring are distributed towards contributors and maintainers based on the contributions they do.
We believe in investing the money towards making webpack better.

But there is a pandemic, and companies ain't that much open to sponsoring anymore.
Webpack is suffering under these circumstances too (like many other companies and people).

We were never able to pay our contributors the amount we think they deserve, but now we only have half of the money available, so we need to make a more serious cut.
Until the situation improves we will only pay contributors and maintainers the first 10 days of each month.

> The remaining days they could work voluntary, paid by their employer, work on something else or take some days off.
> This allows us to pay the work in the first 10 days more equivalent to the invested time.

Reviewer: ```suggestion
The remaining days they could work voluntarily, paid by their employer, work on something else, or take some days off.
This allows us to pay for their work in the first 10 days more equivalent to the invested time.
```

The biggest "Thank You" goes to [trivago](https://tech.trivago.com/opensource) which has been sponsoring webpack a huge amount for the last 3 years.
Sadly they are unable to continue their sponsorship this year, as they have been hit hard by Covid-19.
I hope some other company steps up and follows these (gigantic) footsteps.

Thanks to [all the sponsors](/#sponsors).

## General direction

This release focus on the following:

- We try to improve build performance with Persistent Caching.
- We try to improve Long Term Caching with better algorithms and defaults.
- We try to improve bundle size with better Tree Shaking and Code Generation.

> - We try to improve compatibility to the web platform.

Reviewer: ```suggestion
- We try to improve compatibility with the web platform.
```

- We try to clean up internal structures that were left in a weird state while implementing features in v4 without introducing any breaking changes.

> - We try to prepare for future features by introducing breaking changes now, allowing us to stay on v5 for as long as possible.

Reviewer 2: Remove the "we try to"

All items deprecated in v4 were removed.

MIGRATION: Make sure that your webpack 4 build does not print deprecation warnings.

Here are a few things that were removed but did not have deprecation warnings in v4:

- IgnorePlugin and BannerPlugin must now be passed only one argument that can be an object, string or function.

### Deprecation codes

New deprecations include a deprecation code so they are easier to reference.

### Syntax deprecated

`require.include` has been deprecated and will emit a warning by default when used.

Behavior can be changed with `Rule.parser.requireInclude` to allowed, deprecated or disabled.

### Automatic Node.js Polyfills Removed

In the early days, webpack's aim was to allow running most Node.js modules in the browser, but the module landscape changed and many module uses are now written mainly for frontend purposes. webpack <= 4 ships with polyfills for many of the Node.js core modules, which are automatically applied once a module uses any of the core modules (i.e. the `crypto` module).

While this makes using modules written for Node.js easy, it adds these huge polyfills to the bundle. In many cases these polyfills are unnecessary.

> webpack 5 stops automatically polyfilling these core modules and focus on frontend-compatible modules. Our goal it to improve compatibility with the web platform, where Node.js core modules are not available.

Reviewer: ```suggestion
webpack 5 stops automatically polyfilling these core modules and focus on frontend-compatible modules. Our goal is to improve compatibility with the web platform, where Node.js core modules are not available.
```

In a module like this:

``` js
import { something } from './something';

function usingSomething() {
  return something;
}

export function test() {
  return usingSomething();
}
```

The inner graph algorithm will figure out that `something` is only used when the `test` export is used. This allows to flag more exports as unused and to omit more code from the bundle.

When `"sideEffects": false` is set, this allows to omit even more modules. In this example `./something` will be omitted when the `test` export is unused.

To get the information about unused exports `optimization.unusedExports` is required. To remove side-effect-free modules `optimization.sideEffects` is required.

The following symbols can be analysed:

- function declarations
- class declarations
- `export default` with or variable declarations with
    - function expressions
    - class expressions
    - sequence expressions
    - `/*#__PURE__*/` expressions
    - local variables
    - imported bindings

> FEEDBACK: If you find something missing in this analysis, please report an issue and we consider adding it.

Reviewer 2: Bold this

Many internal plugins will use the Persistent Cache too. Examples: `SourceMapDevToolPlugin` (to cache the SourceMap generation) or `ProgressPlugin` (to cache the number of modules)

The Persistent Cache will automatically create multiple cache files depending on usage to optimize read and write access to and from the cache.

By default timestamps will be used for snapshoting in development mode and file hashes in production mode.
File hashes allow to use Persistent Caching on CI too.

#### Compiler Idle and Close

Compilers now need to be closed after being used. Compilers now enter and leave the idle state and have hooks for these states. Plugins may use these hooks to do unimportant work. (i. e. the Persistent cache slowly stores the cache to disk). On compiler close - All remaining work should be finished as fast as possible. A callback signals the closing as done.

Plugins and their respective authors should expect that some users may forget to close the Compiler. So, all work should eventually be finishing while in idle too. Processes should be prevented from exiting when the work is being done.

> The `webpack()` façade automatically calls `close` when being passed a callback.

Author: proselint enforces `façade`
